scenario/utils.ts

```diff
@@ -254,3 +266,178 @@ export async function fetchLogs(
+) {
+  const MUMBAI_RECEIVER_ADDRESSS = '0x0000000000000000000000000000000000001001';
+  const EVENT_LISTENER_TIMEOUT = 60000;
+
+  const stateSender = await governanceDeploymentManager.getContractOrThrow('stateSender');
+  const timelock = await bridgeDeploymentManager.getContractOrThrow('timelock');
+  const bridgeReceiver = await bridgeDeploymentManager.getContractOrThrow('bridgeReceiver');
+  const fxChild = await bridgeDeploymentManager.getContractOrThrow('fxChild');
+
+  // listen on events on the fxRoot contract
+  const stateSyncedListenerPromise = new Promise(async (resolve, reject) => {
+    const filter = stateSender.filters.StateSynced();
+
+    governanceDeploymentManager.hre.ethers.provider.on(filter, (log) => {
+      resolve(log);
+    });
+
+    setTimeout(() => {
+      reject(new Error('StateSender.StateSynced event listener timed out'));
+    }, EVENT_LISTENER_TIMEOUT);
+  });
+
+  const stateSyncedEvent = await stateSyncedListenerPromise as Event;
+  const { args: { data: stateSyncedData } } = stateSender.interface.parseLog(stateSyncedEvent);
+
+  const mumbaiReceiverSigner = await impersonateAddress(bridgeDeploymentManager, MUMBAI_RECEIVER_ADDRESSS);
+
+  await setNextBaseFeeToZero(bridgeDeploymentManager);
+  const onStateReceiveTxn = await (
+    await fxChild.connect(mumbaiReceiverSigner).onStateReceive(
+      123,             // stateId
+      stateSyncedData, // _data
+      { gasPrice: 0 }
+    )
+  ).wait();
+
+  const queueTransactionEvent = onStateReceiveTxn.events.find(event => event.address === timelock.address);
+  const { args: { target, value, signature, data, eta } } = timelock.interface.parseLog(queueTransactionEvent);
+
+  // fast forward l2 time
+  await setNextBlockTimestamp(bridgeDeploymentManager, eta.toNumber() + 1);
+
+  // execute queued transaction
+  await setNextBaseFeeToZero(bridgeDeploymentManager);
+  await bridgeReceiver.executeTransaction(
+    target,
+    value,
+    signature,
+    data,
+    eta,
+    { gasPrice: 0 }
+  );
+}
+
+export async function relayMessage(
+  governanceDeploymentManager: DeploymentManager,
+  bridgeDeploymentManager: DeploymentManager
+) {
+  const bridgeNetwork = bridgeDeploymentManager.network;
+  switch (bridgeNetwork) {
+    case 'mumbai':
+      await relayMumbaiMessage(governanceDeploymentManager, bridgeDeploymentManager);
+      break;
+    default:
+      throw new Error(`No governance execution strategy for network: ${bridgeNetwork}`);
+  }
+}
+
+export async function fastL2GovernanceExecute(
+  governanceDeploymentManager: DeploymentManager,
+  bridgeDeploymentManager: DeploymentManager,
+  proposer: SignerWithAddress,
+  targets: string[],
+  values: BigNumberish[],
+  signatures: string[],
+  calldatas: string[]
+) {
+  await fastGovernanceExecute(
+    governanceDeploymentManager,
+    proposer,
+    targets,
+    values,
+    signatures,
+    calldatas
+  );
+
+  await relayMessage(governanceDeploymentManager, bridgeDeploymentManager);
```

Comment:
Wouldn't `relayMessage` have to be called before `fastGovernanceExecute` to initiate the event listener on `fxRoot`?

Reply:
synced offline; yeah, you'd think so, but apparently Ethers does some retroactive searching for events because setting the listener after the event has already occurred still returns the desired event 🤷